Desktop: Fixes #10007: Fixed Toggle Comment & Delete/Duplicate/Sort Line Options in Beta Editor

criticic opened the pull request:
<!--

Please prefix the title with the platform you are targetting:

Here are some examples of good titles:

- Desktop: Resolves #123: Added new setting to change font
- Mobile, Desktop: Fixes #456: Fixed config screen error
- All: Resolves #777: Made synchronisation faster

And here's an explanation of the title format:

- "Desktop" for the Windows/macOS/Linux app (Electron app)
- "Mobile" for the mobile app (or "Android" / "iOS" if the pull request only applies to one of the mobile platforms)
- "CLI" for the CLI app

If it's two platforms, separate them with commas - "Desktop, Mobile" or if it's for all platforms, prefix with "All".

If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform.

Then please append the issue that you've addressed or fixed. Use "Resolves #123" for new features or improvements and "Fixes #123" for bug fixes.

AND PLEASE READ THE GUIDE: https://github.com/laurent22/joplin/blob/dev/readme/dev/index.md

-->

@PackElend label me please, Fixes #10007

- Fixes the Delete Line & Duplicate Line & Toggle Comment Options in the Beta Editor
- Added a test to verify the behaviour

laurent22 commented on packages/editor/CodeMirror/CodeMirrorControl.test.ts:
```diff
@@ -57,4 +57,23 @@ describe('CodeMirrorControl', () => {
 		expect(command).toHaveBeenCalledTimes(1);
 	});
+
+	it('toggleComment should work', () => {
```
Please use the same pattern as other tests - `it('should...`. And generally we should use plain English in the description so something `it('should toggle comments'...`

laurent22 commented on packages/editor/CodeMirror/CodeMirrorControl.test.ts:
```diff
@@ -57,4 +57,23 @@ describe('CodeMirrorControl', () => {
+		expect(control.getValue()).toBe('Hello\nWorld\n');
+	});
+
+	it('deleteLine should work', () => {
```
Same here

criticic replied:
updated it!

criticic commented:
I also added the duplicate line option in the v6 editor. It required a custom EditorCommandFunction, as far as I have tested the v5 version, it has the same behaviour.

Also added tests for it

personalizedrefrigerator commented on packages/editor/CodeMirror/editorCommands/duplicateLine.ts:
```diff
@@ -0,0 +1,32 @@
+			insertPos = range.to;
+		}
+		return {
+			range: EditorSelection.range(range.to, range.to + text.length),
```
In the CodeMirror 5 editor, it seems to leave the selection as just a single cursor when there is no selection. It might make sense to duplicate that behavior here.

criticic replied:
done.

criticic commented:
Also added the sort selected lines option in v6 editor, and its corresponding EditorCommandFunction. It is based on the old v5 plugin present at https://github.com/laurent22/joplin/blob/dev/packages/app-desktop/gui/NoteEditor/NoteBody/CodeMirror/utils/useLineSorting.ts

Also added tests.

personalizedrefrigerator commented on packages/editor/types.ts:
```diff
@@ -37,6 +39,7 @@ export enum EditorCommandType {
 
 	// Editing and navigation commands
 	ScrollSelectionIntoView = 'scrollSelectionIntoView',
+	DeleteLine = 'deleteLine',
```
For me, <kbd>ctrl</kbd>-<kbd>d</kbd> continues to select the current word, rather than deleting the current line. (I'm on Linux).

I suspect this is related to the default CodeMirror 6 keybindings (search for `...searchKeymap` in `createEditor.ts`).

personalizedrefrigerator replied:
This does seem to be caused by `searchKeymap`. Removing `searchKeymap` fixes this issue, but it also breaks several search keybindings (e.g. <kbd>Esc</kbd> to close the search dialog).

Because it's a different issue, removing `searchKeymap` or fixing this issue in another way should probably be done in a different pull request.

criticic replied:
yeah will fix it in a separate pr then.
though Edit > Delete Line works.

personalizedrefrigerator replied:
I've created an issue for it: https://github.com/laurent22/joplin/issues/10020

personalizedrefrigerator commented on packages/editor/CodeMirror/CodeMirrorControl.test.ts:
```diff
@@ -57,4 +57,47 @@ describe('CodeMirrorControl', () => {
+		expect(control.getValue()).toBe('Hello\n');
+	});
+
+	it('should duplicate line', () => {
```
While tests for duplicate and sort could make sense here, it might make sense to move these tests to new `duplicateLine.test.ts` and `sortSelectedLines.test.ts` files.

This could be better because:
- It would be easier to find tests specific to these commands.
- If these commands are eventually disconnected from `editorCommands.ts` (e.g. because of a refactoring), their tests will continue to work.

It could also have the following drawbacks:
- It probably wouldn't make sense to construct a `CodeMirrorControl` in such files (if we want them to pass independent of which commands are in `editorCommands.ts`).

criticic replied:
should I keep these tests here then? or would checking the behaviour the .test.ts only be enough?

personalizedrefrigerator replied:
I suggest moving tests for `duplicateLine` and `sortSelectedLines` to `duplicateLine.test.ts` and `sortSelectedLines.test.ts` files.

criticic commented:
so noticed that swapLine also did not have a test so added that as well, have made all tests independent of editorCommands as suggested

personalizedrefrigerator commented on packages/editor/CodeMirror/editorCommands/duplicateLine.ts:
```diff
@@ -0,0 +1,34 @@
+	const state = editor.state;
+	const doc = state.doc;
+
+	const transaction = state.changeByRange(range => {
```
Currently, this duplicates selected lines multiple times if there are multiple cursors on a line. However, because the CodeMirror 5 editor has the same behavior, I think it's fine to leave this as-is.

personalizedrefrigerator commented on packages/editor/CodeMirror/editorCommands/swapLine.test.ts:
```diff
@@ -0,0 +1,32 @@
+import swapLine from './swapLine';
+
+
+describe('swapLine', () => {
```
Thank you for adding this!

personalizedrefrigerator commented on packages/editor/CodeMirror/editorCommands/swapLine.test.ts:
```diff
@@ -0,0 +1,32 @@
+			selection: EditorSelection.cursor(0),
+		});
+
+		swapLine(1)(editorView);
```
```suggestion
		swapLine(SwapLineDirection.Down)(editorView);
```

Please use the `enum` here.

criticic replied:
updated it!

personalizedrefrigerator commented on packages/editor/CodeMirror/editorCommands/swapLine.test.ts:
```diff
@@ -0,0 +1,32 @@
+			selection: EditorSelection.cursor(6),
+		});
+
+		swapLine(-1)(editorView);
```
```suggestion
		swapLine(SwapLineDirection.Up)(editorView);
```

personalizedrefrigerator reviewed (commented):
I've left one minor stylistic comment. Aside from that, this looks good to me!

I've tested it locally on Linux (OpenSUSE) and the commands all seem to work.

Thank you for working on this!
